src/Mvc/Mvc.Core/src/Infrastructure/ResourceInvoker.cs

```diff
@@ -57,46 +57,139 @@ internal abstract class ResourceInvoker
             _cursor = new FilterCursor(filters);
         }
 
-        public virtual async Task InvokeAsync()
+        public virtual Task InvokeAsync()
         {
+            if (_diagnosticListener.IsEnabled() || _logger.IsEnabled(LogLevel.Information))
+            {
+                return Logged(this);
+            }
+
+            _actionContextAccessor.ActionContext = _actionContext;
+            var scope = _logger.ActionScope(_actionContext.ActionDescriptor);
+
+            Exception invokeException = null;
+            Task task = null;
             try
             {
-                _actionContextAccessor.ActionContext = _actionContext;
+                task = InvokeFilterPipelineAsync();
+            }
+            catch (Exception ex)
+            {
+                invokeException = ex;
+            }
 
-                _diagnosticListener.BeforeAction(
-                    _actionContext.ActionDescriptor,
-                    _actionContext.HttpContext,
-                    _actionContext.RouteData);
+            if (invokeException != null)
+            {
+                return Awaited(this, Task.FromException(invokeException), scope);
+            }
```

Comment:
Is there a reason not to fold this into the `catch` block?

Reply:
Is this pattern going to negatively effect the quality of callstacks?

Reply:
Good point. Its a vestige of the multiple iterations I went through to deal with the double finally block

Reply:
> Is this pattern going to negatively effect the quality of callstacks?

Yes; because local function names become garbage and there is no contract with the C# compiler that the runtime can rely on to turn them back into anything sensible https://github.com/dotnet/roslyn/issues/22857. This came up when turning async callstacks into something nice https://github.com/dotnet/coreclr/pull/14652 /cc @tmat @danmosemsft 

I can move them out into regular static methods, and they will show up sensibility if preferred?

Reply:
I'm more concerned with whether callstacks get truncated. Do you have an example of what thowing an exception from a controller looks like (developer exception page) from your current PR?

Reply:
Interestingly its the same for
```csharp
throw new Exception();
```
and
```csharp
await Task.Delay(100);
throw new Exception();
```

![image](https://user-images.githubusercontent.com/1142958/56073307-7b359780-5d99-11e9-9fd1-e9c56e92f2dc.png)

Whereas I thought the second one would look terribad.

Reply:
In a page, again the delayed and immediate produce the same stack traces which is
![image](https://user-images.githubusercontent.com/1142958/56073360-5a217680-5d9a-11e9-93a5-14fb23a76123.png)

Reply:
the other version

![image](https://user-images.githubusercontent.com/1142958/56073371-85a46100-5d9a-11e9-9be1-eedc18a720e3.png)